src/backend/error.rs

```diff
@@ -0,0 +1,242 @@
+use std::{error::Error, fmt::Display};
+
+use super::Context;
+
+pub(crate) fn format_error(err: &(impl Error + 'static), context: &Context) -> String {
+    let mut err_descs = Vec::new();
+
+    err_descs.push(fmt_pretty_any(err, context));
+
+    let mut source_opt = err.source();
+    while let Some(source) = source_opt {
+        err_descs.push(fmt_pretty_any(source, context));
+        source_opt = source.source();
+    }
+
+    let desc = format!("Validation Error\n\nCaused by:\n{}", err_descs.join(""));
+    desc
+}
+
+fn fmt_pretty_any(error: &(dyn Error + 'static), context: &Context) -> String {
+    if let Some(pretty_err) = error.downcast_ref::<super::direct::ContextError>() {
```

Comment:
can this be casted to a `PrettyError` trait object?

Reply:
Needs to be a concrete type
> the `downcast_ref` method cannot be invoked on a trait object